### DISCUSSION TITLE
v0.10.1

### DISCUSSION THREAD

**Discussion by laipz8200** [langgenius/dify#9742] [Releases]
## 🚀 What’s New in v0.10.1? 🚀

Welcome to version v0.10.1, where we've made significant strides in expanding functionalities and enhancing user workflows. Here's a quick look at the highlights of this release:

### 🔹 **Claude 3.5 Sonnet v2 Models**

Experience advanced language processing across Google Cloud, Bedrock, and Anthropic platforms, seamlessly integrated with the latest technology advancements via [#9688](https://github.com/langgenius/dify/pull/9688), [#9685](https://github.com/langgenius/dify/pull/9685), and [#9708](https://github.com/langgenius/dify/pull/9708).

### ✨ **Code Generator**

Empower your Workflow's Code Node with our new Code Generator tool. This feature is designed to streamline coding processes within workflows, boosting development efficiency and reducing manual coding efforts. Check it out in [#9051](https://github.com/langgenius/dify/pull/9051).

### 🐛 Bug Fixes

- **Custom Icon Restoration**: We've reintroduced the flexibility and personalization of custom icons, addressing previous issues to ensure your icons display correctly, as implemented by @zxhlyh in [#9654](https://github.com/langgenius/dify/pull/9654).

- **HTTP Node Improvements**: Enhanced JSON handling and simplified request management to boost the stability and security of the HTTP Node by @laipz8200 in [#9616](https://github.com/langgenius/dify/pull/9616), and fixed the handling of empty and string data inputs for consistent performance by @laipz8200 in [#9579](https://github.com/langgenius/dify/pull/9579).

- **File Upload Corrections**: Resolved multiple issues affecting file uploads for smoother operations. Image config uploads were corrected by @laipz8200 in [#9568](https://github.com/langgenius/dify/pull/9568).

- **Improved Iteration Handling**: Made iterators more robust by gracefully handling empty scenarios, as addressed by @laipz8200 in [#9565](https://github.com/langgenius/dify/pull/9565).

- **VisionConfig Validation**: Introduced a validator for `VisionConfig` entities to handle `None` values properly, improving data integrity by @laipz8200 in [#9598](https://github.com/langgenius/dify/pull/9598).

- **Workflow Management**: Improved workflow and database session management to resolve instabilities and enhance reliability by @laipz8200 in [#9581](https://github.com/langgenius/dify/pull/9581).

- **If-Else Node Enhancement**: Enhanced if-else nodes to properly add downstream nodes, ensuring logical consistency in workflow paths by @linfey90 in [#9640](https://github.com/langgenius/dify/pull/9640).

- **Used File in Tools**: Addressed errors caused by file operations in tools to ensure smoother functionality by @iamjoel in [#9660](https://github.com/langgenius/dify/pull/9660).

- **Environment Variables**: Fixed environmental variable issues for ModelProvider and Tool Position to ensure they operate as intended. Resolved by @tkoshida in [#9650](https://github.com/langgenius/dify/pull/9650).

- **Web App UX**: Corrected the style of the features panel specifically in Safari to enhance the overall visual experience by @JzoNgKVO in [#9573](https://github.com/langgenius/dify/pull/9573) and addressed the email login redirect issue and corrected navigation behaviors to improve user experience by @ZhouhaoJiang in [#9698](https://github.com/langgenius/dify/pull/9698).

This update is packed with critical fixes, new features, and exciting enhancements to empower your AI applications. Dive into v0.10.1 and experience these improvements firsthand!

---

### Docker compose deployments

1. Back up your customized docker-compose YAML file (optional)

   ```bash
   cd docker
   cp docker-compose.yaml docker-compose.yaml.$(date +%s).bak
   ```

2. Get the latest code from the main branch

   ```bash
   git checkout main
   git pull origin main
   ```

3. Stop the service，Command, please execute in the docker directory

   ```bash
   docker compose down
   ```

4. Back up data

   ```bash
   tar -cvf volumes-$(date +%s).tgz volumes
   ```

5. Upgrade services

   ```bash
   docker compose up -d
   ```

### Source Code deployments

1. Stop API server, Worker and Web frontend Server.

2. Get the latest code from the release branch:

   ```bash
   git checkout 0.10.1
   ```

3. Update Python dependencies:

   ```bash
   cd api
   poetry install
   ```

7. Then, let's run the migration script:

   ```bash
   poetry run flask db upgrade
   ```

8. Finally, run API server, Worker and Web frontend Server again.

---

## What's Changed
* Build/fix wrong icon name by @AkaraChen in https://github.com/langgenius/dify/pull/9527
* enhance: use urllib join instead of fstring by @41tair in https://github.com/langgenius/dify/pull/9549
* feat: Update tongyi models by @AAEE86 in https://github.com/langgenius/dify/pull/9552
* chore: lint code to remove unused imports and variables by @DDDDD12138 in https://github.com/langgenius/dify/pull/9553
* chore: format get_customizable_model_schema return value by @yaoice in https://github.com/langgenius/dify/pull/9335
* fix(upload): correct incorrect dictionary key usage by @laipz8200 in https://github.com/langgenius/dify/pull/9563
* Fix: file upload support extension .md by @JzoNgKVO in https://github.com/langgenius/dify/pull/9564
* fix(iteration): handle empty iterator gracefully by @laipz8200 in https://github.com/langgenius/dify/pull/9565
* fix(file upload): correct upload method key for image config by @laipz8200 in https://github.com/langgenius/dify/pull/9568
* Fix: style of features panel in safari by @JzoNgKVO in https://github.com/langgenius/dify/pull/9573
* dep: bump pydantic to 2.9 by @bowenliang123 in https://github.com/langgenius/dify/pull/9077
* fix(validation): improve variable handling and validation by @laipz8200 in https://github.com/langgenius/dify/pull/9578
* fix(http_request): handle empty and string data inputs by @laipz8200 in https://github.com/langgenius/dify/pull/9579
* fix(workflow): improve database session handling and variable management by @laipz8200 in https://github.com/langgenius/dify/pull/9581
* fix(workflow): handle NoneSegments in variable extraction by @laipz8200 in https://github.com/langgenius/dify/pull/9585
* refactor(template_transform): use keyword-only arguments by @laipz8200 in https://github.com/langgenius/dify/pull/9575
* refactor: add logging extension module for log initialization by @hwzhuhao in https://github.com/langgenius/dify/pull/9524
* chore: cleanup ineffective linter rules exclusions by @bowenliang123 in https://github.com/langgenius/dify/pull/9580
* fix(entities): add validator for `VisionConfig` to handle None values by @laipz8200 in https://github.com/langgenius/dify/pull/9598
* refactor(variables): replace deprecated 'get_any' with 'get' method by @laipz8200 in https://github.com/langgenius/dify/pull/9584
* refactor: use dify_config to replace legacy usage of flask app's config by @bowenliang123 in https://github.com/langgenius/dify/pull/9089
* chore: (#9089 followup) fix storage factory constructor by @bowenliang123 in https://github.com/langgenius/dify/pull/9609
* fix(core): use CreatedByRole enum for role consistency by @laipz8200 in https://github.com/langgenius/dify/pull/9607
* fix(storage): use centralized config management by @laipz8200 in https://github.com/langgenius/dify/pull/9620
* fix(storage): ensure `storage_runner` initialization within app context by @laipz8200 in https://github.com/langgenius/dify/pull/9627
* fix: wrong url of guides doc in new feature panel by @hanfangyuan4396 in https://github.com/langgenius/dify/pull/9626
* Fix: doc link of legacy features by @JzoNgKVO in https://github.com/langgenius/dify/pull/9634
* fix: update the default model to gpt-4o-mini for duckduckgo ai chat by @hwzhuhao in https://github.com/langgenius/dify/pull/9614
* fix(http_request): simplify JSON handling in requests by @laipz8200 in https://github.com/langgenius/dify/pull/9616
* fix: fields.Nested(message_file_fields) by @crazywoola in https://github.com/langgenius/dify/pull/9632
* fix(api): enhance file factory URL handling by @laipz8200 in https://github.com/langgenius/dify/pull/9631
* fix: remove url join by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/9635
* fix: file type document is not supported by @hjlarry in https://github.com/langgenius/dify/pull/9618
* refactor(api): simplify limit retrieval and return types by @laipz8200 in https://github.com/langgenius/dify/pull/9641
* fix(files): update Content-Length handling for tool and remote files by @laipz8200 in https://github.com/langgenius/dify/pull/9649
* fix(segments): return empty string instead of "null" for text, log, and markdown properties by @laipz8200 in https://github.com/langgenius/dify/pull/9651
* Fix: type missing of remote file in chat by @JzoNgKVO in https://github.com/langgenius/dify/pull/9652
* fix: webapp answer icon by @zxhlyh in https://github.com/langgenius/dify/pull/9654
* fix: tool use file caused error by @iamjoel in https://github.com/langgenius/dify/pull/9660
* Fix/retrieval setting weight default value by @zxhlyh in https://github.com/langgenius/dify/pull/9622
* fix: environment variables for ModelProvider and Tool Position are not working by @tkoshida in https://github.com/langgenius/dify/pull/9650
* Web app support sending message using numpad enter by @xuzuodong in https://github.com/langgenius/dify/pull/9659
* feat(Tools): Refactor the base table plugin by @hgnulb in https://github.com/langgenius/dify/pull/9182
* chore(models): convert created_by_role to its value for consistency by @laipz8200 in https://github.com/langgenius/dify/pull/9612
* feat: add code generator by @Kota-Yamaguchi in https://github.com/langgenius/dify/pull/9051
* fix: add downstream nodes of this branch by @linfey90 in https://github.com/langgenius/dify/pull/9640
* feat: add upstash as a new vector database provider by @userzhangqg in https://github.com/langgenius/dify/pull/9644
* style: chat answer align with new UI by @xuzuodong in https://github.com/langgenius/dify/pull/9658
* fix(variable_pool): handle invalid attributes in variable lookup by @laipz8200 in https://github.com/langgenius/dify/pull/9646
* chore: reuse existing test functions with upstash vdb by @yaoice in https://github.com/langgenius/dify/pull/9679
* Feat/optimize login by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/9642
* feat(api): enhance file preview handling by @laipz8200 in https://github.com/langgenius/dify/pull/9674
* feat(condition): add support for 'exists' and 'not exists' operators by @laipz8200 in https://github.com/langgenius/dify/pull/9687
* add bedrock claude 3.5 v2 support by @ybalbert001 in https://github.com/langgenius/dify/pull/9685
* fix: remove email code login redirect by @ZhouhaoJiang in https://github.com/langgenius/dify/pull/9698
* feat(api): add video and audio file size limits to upload config by @laipz8200 in https://github.com/langgenius/dify/pull/9703
* fix: workflow [if node] checklist by @zxhlyh in https://github.com/langgenius/dify/pull/9699
* Help documentation URL correction by @AAEE86 in https://github.com/langgenius/dify/pull/9704
* fix: revert ref usage in handleFormChange to fix IME input issues by @kurokobo in https://github.com/langgenius/dify/pull/9672
* feat: added claude 3.5 sonnet v2 model to Google Cloud Vertex AI by @hellof20 in https://github.com/langgenius/dify/pull/9688
* nltk security issue and upgrade unstructured by @JohnJyong in https://github.com/langgenius/dify/pull/9558
* build(deps): bump mermaid from 10.4.0 to 10.9.3 in /web by @dependabot in https://github.com/langgenius/dify/pull/9709
* feat(model_runtime): add new model 'claude-3-5-sonnet-20241022' by @laipz8200 in https://github.com/langgenius/dify/pull/9708
* Feat: use file size limit from api by @JzoNgKVO in https://github.com/langgenius/dify/pull/9711
* build(deps): bump next from 14.2.4 to 14.2.10 in /web by @dependabot in https://github.com/langgenius/dify/pull/9713
* feat: support user-defined configuration of log file size and retention count by @hwzhuhao in https://github.com/langgenius/dify/pull/9610
* Revert "Feat: use file size limit from api" by @JzoNgKVO in https://github.com/langgenius/dify/pull/9714
* fix: refresh current page if url contains token by @douxc in https://github.com/langgenius/dify/pull/9718
* chore: update version to 0.10.1 by @laipz8200 in https://github.com/langgenius/dify/pull/9689
* Modify characters by @AAEE86 in https://github.com/langgenius/dify/pull/9707
* remove ppt import by @JohnJyong in https://github.com/langgenius/dify/pull/9721
* downgrade unstructured nltk version by @JohnJyong in https://github.com/langgenius/dify/pull/9726
* add vdb py test by @JohnJyong in https://github.com/langgenius/dify/pull/9706
* fixed: variable reference error by @luckylhb90 in https://github.com/langgenius/dify/pull/9722

## New Contributors
* @AkaraChen made their first contribution in https://github.com/langgenius/dify/pull/9527
* @41tair made their first contribution in https://github.com/langgenius/dify/pull/9549
* @tkoshida made their first contribution in https://github.com/langgenius/dify/pull/9650
* @linfey90 made their first contribution in https://github.com/langgenius/dify/pull/9640

**Full Changelog**: https://github.com/langgenius/dify/compare/0.10.0...0.10.1

<hr /><em>This discussion was created from the release <a href='https://github.com/langgenius/dify/releases/tag/0.10.1'>v0.10.1</a>.</em>

**Comment by zhangyifei1** [comment]
源码启动api服务时间很慢，是什么原因？

**Reply by zhangyifei1** [reply]
> 启动api服务时间很慢，源码是什么原因？
一直停留这个地方二十多分钟
![Uploading image.png…]()

**Reply by AAEE86** [reply]
请重新上传图片

**Comment by nadirvishun** [comment]
The `dockerfile` compile prompted that `libldap-2.5-0=2.5.18+dfsg-3` and `perl=5.38.2-5` were not found, and then went to [deban testing](https://www.debian.org/distrib/packages) and found that the only versions were ` libldap-2.5-0=2.5.18+dfsg-3+b1` and `perl=5.40.0-6`, I don't know if anyone else has also encountered this situation

**Reply by laipz8200** [reply]
Dify uses some newer versions of libraries, and their updates may cause this problem. If you find that the build fails, you can provide us with a PR containing the latest version number and the corresponding description page.

**Reply by laipz8200** [reply]
#9801
